### DISCUSSION TITLE
How to account for edges with Mitsuba?

### DISCUSSION THREAD

**Discussion by jeertmans** [mitsuba-renderer/mitsuba3#717] [Q&A]
Hello!

As part of some Ray Tracing project, I'd like to add diffraction, for which I need edges.

Currently, I don't see any way for obtaining edges (on which diffraction can occur). I tried to manually incorporate them in the `.ply` files, but they are ignored:

![image](https://user-images.githubusercontent.com/27275099/236251608-0a957d11-6dc6-4a04-be43-1e3057b619c9.png)

Is there a workaround for that?

Thanks!

**Comment by njroussel** [comment]
Hi @jeertmans 

The `ply` plugin only supports attributes on faces and vertices currently, hence the error message.

I'm not quite sure how I would go about implementing diffraction. I'm assuming you only want to consider a subsert of the shape's edges - in which case you'll need to modify the `Mesh` class to support edge attributes. Another approach would be to use face attributes and store something like a binary triplet to encode which of it's edges should be considered for diffraction.
Not quite sure what's the easiest/best solution for you.

**Reply by jeertmans** [reply]
Hi @njroussel, thanks for your answer :-)

Well, implementing diffraction will be done with one method I developed, but I only need diffraction candidates (so "exterior" edges).

As post-processing a mesh structure to obtain those edges is quite challenging (many vertices are duplicates, neighboring triangles introduce duplicated edges, co-planar edges must be discarded, ...), I would rather create a small-scale geometry by hand, and manually tell what are the "edges" I would like to consider.

Another solution is that I don't rely on .PLY file, but try to use the .OBJ or serialized formats. I should try this :-)

**Comment by ghost** [comment]
Hi, @jeertmans!

I am also interested in diffraction. I wonder how you could simulate it with a ray tracer. Could you give me some hints?

**Reply by jeertmans** [reply]
Hi @nick008a, it really depends if you are doing ray launching or ray tracing. It's a bit of self-promotion, but I wrote a paper about simulating diffraction (and others) on a radio propagation context: https://arxiv.org/abs/2301.06399. This is a very general (and not fast) method, but other exist, like minimizing the path length (Fermat's principle).

**Reply by ghost** [reply]
Thanks for the reply! Is this possible in mitsuba framework? Or do you have any good code base to get started?

**Reply by jeertmans** [reply]
With Mitsuba, you should implement all the diffraction from scratch. If you are into radio propagation, you can take a look at Sionna, which implements very basic diffraction. I am myself implementing my own framework, but it is not ready to use yet.
